lab_5_scrapper/scrapper.py

```diff
@@ -204,12 +463,78 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
     Args:
         base_path (Union[pathlib.Path, str]): Path where articles stores
     """
+    if base_path.exists():
+        shutil.rmtree(base_path.parent)
+    base_path.mkdir(parents=True)
+
+
+class ProgramKilledException(Exception):
+    """
+    Raise when the program is killed.
+    """
+
+
+class Killer:
+    """
+    Class to see if the system is shut down or script is interrupted.
+    """
+    def __init__(self) -> None:
+        signals = [signal.SIGINT, signal.SIGTERM, signal.SIGBREAK]
+        for sig in signals:
+            signal.signal(sig, self.handler)
+
+    def handler(self, signum: int, frame: FrameType | None) -> None:
+        raise ProgramKilledException
+
+
+def save_all_urls(crawler: CrawlerRecursive) -> None:
+    with open('already_crawled_urls.txt', 'w', encoding='utf-8', newline='\n') as f:
+        f.write("\n".join(crawler.already_crawled))
 
 
 def main() -> None:
     """
     Entrypoint for scrapper module.
     """
+    configuration = Config(path_to_config=constants.CRAWLER_CONFIG_PATH)
+
+    prepare_environment(base_path=constants.ASSETS_PATH)
+
+    crawler = Crawler(config=configuration)
+    crawler.find_articles()
+
+    for index, url in enumerate(crawler.urls):
+        parser = HTMLParser(full_url=url, article_id=index + 1, config=configuration)
+        article = parser.parse()
+        if isinstance(article, Article):
+            to_raw(article)
+            to_meta(article)
+
+
+def main_recursive() -> None:
+    """
+    Entrypoint for recursive scrapper module.
+    """
+    configuration = Config(path_to_config=constants.CRAWLER_CONFIG_PATH)
+
+    prepare_environment(constants.ASSETS_PATH)
+    crawler = CrawlerRecursive(config=configuration)
+
+    try:
+        Killer()
+        crawler.find_articles()
+    except ProgramKilledException:
+        save_all_urls(crawler)
```

Comment:
you plan to be quick on saving based on signal sent to a program. OK. but now imagine - your computer is halted after the crash of electricity station in your area. do you agree there will not be such nice signal propagation to all running programs that politely asks to finish? rework to that case

Reply:
I added the continuous addition of the URLs so that if anything happens, the URLs are already collected to the point of where I am right now